test/docker-e2e/e2e_upgrade_test.go

```diff
@@ -0,0 +1,221 @@
+		s.Require().Equal(uint32(0), resp.Code, "signal tx failed with code %d", resp.Code)
+	}
+
+	// wait a block so the signals are included
+	s.Require().NoError(wait.ForBlocks(ctx, 1, chain))
+
+	s.validateSignalTally(ctx, validatorNode, targetAppVersion)
+
+	msgTry := signaltypes.NewMsgTryUpgrade(txClient.DefaultAddress())
+	resp, err := txClient.SubmitTx(ctx, []sdk.Msg{msgTry}, user.SetGasLimit(gasLimit), user.SetFee(fee))
+	s.Require().NoError(err, "failed to broadcast try-upgrade tx")
+	s.Require().Equal(uint32(0), resp.Code, "try-upgrade tx failed with code %d", resp.Code)
+
+	// Wait for one block so that the upgrade transaction is processed
+	s.Require().NoError(wait.ForBlocks(ctx, 1, chain))
+
+	// Query upgrade info via gRPC
+	s.T().Log("Querying upgrade info via gRPC")
+	client, cleanup, err := getSignalQueryClient(ctx, validatorNode)
+	s.Require().NoError(err)
+	defer cleanup()
+
+	upgradeResp, err := client.GetUpgrade(ctx, &signaltypes.QueryGetUpgradeRequest{})
+	s.Require().NoError(err, "failed to query upgrade info via gRPC")
+
+	// Ensure an upgrade is indeed pending
+	s.Require().NotNil(upgradeResp.Upgrade, "no upgrade pending after try-upgrade")
+
+	upgradeHeight := upgradeResp.Upgrade.UpgradeHeight
+	s.T().Logf("Upgrade info: app_version=%d height=%d", upgradeResp.Upgrade.AppVersion, upgradeHeight)
+
+	return upgradeHeight
+}
+
+// validateSignalTally queries the signal tally for the given app version and verifies
+// that the voting power meets or exceeds the threshold power.
+func (s *CelestiaTestSuite) validateSignalTally(ctx context.Context, node tastoratypes.ChainNode, appVersion uint64) {
+	s.T().Logf("Querying signal tally for app version %d", appVersion)
+
+	client, cleanup, err := getSignalQueryClient(ctx, node)
+	s.Require().NoError(err)
+	defer cleanup()
+
+	resp, err := client.VersionTally(ctx, &signaltypes.QueryVersionTallyRequest{Version: appVersion})
+	s.Require().NoError(err, "failed to query tally")
+
+	// Verify that voting power meets or exceeds threshold
+	s.Require().True(resp.VotingPower >= resp.ThresholdPower, "voting power (%d) does not meet threshold (%d)", resp.VotingPower, resp.ThresholdPower)
+}
+
+// getSignalQueryClient returns a signaltypes.QueryClient for the provided node.
+// If the node already exposes a live *grpc.ClientConn (docker ChainNode), that
+// connection is reused and the returned cleanup is a no-op. Otherwise the
+// helper dials the node’s gRPC endpoint (port 9090) and returns a cleanup
+// function that closes the connection.
+func getSignalQueryClient(ctx context.Context, node tastoratypes.ChainNode) (signaltypes.QueryClient, func(), error) {
+	if dcNode, ok := node.(*tastoradockertypes.ChainNode); ok && dcNode.GrpcConn != nil {
+		return signaltypes.NewQueryClient(dcNode.GrpcConn), func() {}, nil
+	}
+	host, err := node.GetInternalHostName(ctx)
+	if err != nil {
+		return nil, nil, err
+	}
+	endpoint := fmt.Sprintf("%s:9090", host)
```

Comment:
this won't be resolvable from within the test itself, port `9090` is exposed within the docker network. Since the test is not running inside the network this won't work.

We would need to also expose the host mapping port if we want to do this fallback logic.

Reply:
Sorry I don't get your point. The tests is working and it can query the upgrade signal status right now. What is the issue here? do we need to use IP or under some circumstance, it wont work?

Reply:
ok got it. let me remove that fallback so we wont have random bug in future

Reply:
removed